### PR TITLE
[expression] Language support for format_date() and to_{date,datetime,time}()

### PR DESCRIPTION
**Pull request by nirvn**
## Description

By default, those expression use the application's locale. The addition of an optional language parameter allows handling of dates that wouldn't match that default locale (say for e.g. an English system running QGIS trying to transform a French-formatted string into a date object).

This addresses #32027 (which was a feature request/issue hybrid really).

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by nirvn** [comment]
For the record, there's an argument to be made for this commit to be backported to 3.10 as it fixes the default locale not respecting that of QGIS' app wise locale. The fix for that is the move away from QDate::fromString to QLocale::fromDate. The optional language parameter is merely a cherry on top of the fix.

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -1025,16 +1025,34 @@ static QVariant fcnToString( const QVariantList &values, const QgsExpressionCont
 static QVariant fcnToDateTime( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
 {
   QString format = QgsExpressionUtils::getStringValue( values.at( 1 ), parent );
-  if ( format.isEmpty() )
+  QString language = QgsExpressionUtils::getStringValue( values.at( 2 ), parent );
+  if ( format.isEmpty() && !language.isEmpty() )
+  {
+    parent->setEvalErrorString( QObject::tr( "A format is required to conver to DateTime when the language is specified" ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp]: "convert"

File: src/core/expression/qgsexpressionfunction.cpp [hunk 2]
```diff
@@ -1025,16 +1025,34 @@ static QVariant fcnToString( const QVariantList &values, const QgsExpressionCont
...
+    return QVariant( QDateTime() );
+  }
+
+  if ( format.isEmpty() && language.isEmpty() )
     return QVariant( QgsExpressionUtils::getDateTimeValue( values.at( 0 ), parent ) );
 
   QString datetimestring = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
-  QDateTime datetime = QDateTime::fromString( datetimestring, format );
+  QLocale *locale;
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp]: This doesn't need to be a pointer, does it?

**Comment by nirvn** [comment]
@nyalldawson , comments addressed, good to merge? Any strong feeling on whether to backport this to 3.10 or not?

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
>  Any strong feeling on whether to backport this to 3.10 or not?

I'm -0, but then I've never been personally been affected by locale issues and have no idea on the impact of this.

**Comment by nirvn** [comment]
@nyalldawson , K, thanks for your comment. The more I think about it, the more I feel this needs backporting. The to_date()/to_datetime()/to_time() functions are too unreliable otherwise.

**Pull request merged**